File: algorithms_and_data_structure/with_kotlin/algoritms/LinearSearchAlgorithms.kt
/**
 * Idea: Go through the vector, analyzing each position i. If A[i].key equals K, then it found the
 * position and returns i. If you went through the entire vector and no element equal to K exists,
 * then return -1.
 *
 * Linear search is an algorithm which finds the position of a target value within an array (Usually
 * unsorted)
 *
 * Worst-case performance O(n) Best-case performance O(1) Average performance O(n) Worst-case space
 * complexity O(1)
 */
fun main() {
    val array = intArrayOf(-6, -1, 3, 7, 10, 27, 35, 37, 52)
    val k1 = 27
    val k2 = 2
    println("Result for k1 ${linearSearch(array,  k1)}")
    println("Result for k2 ${linearSearch(array,  k2)}")
}

/**
 * @param array is an array where the element should be found
 * @param k is an element which should be found
 * @return index of the element
 */
fun linearSearch(array: IntArray, k: Int): Int {
    for (i in array.indices) {
        if (array[i].compareTo(k) == 0) {
            return i
        }
    }

    return -1
}

File: algorithms_and_data_structure/dart/bin/algorithms/linear_search_algorithms.dart
/// Idea: Go through the vector, analyzing each position i.
/// If A[i].key equals K, then it found the position and returns i.
/// If you went through the entire vector and no element equal to K exists,
/// then return -1.

void main() {
  final array = [3, -10, 30, 7, 100, 27, 3, 7, 52];
  var k1 = 27;
  var k2 = -1;
  var n = array.length;
  print('Result for k1 ${linearSearch(array, n, k1)}');
  print('Result for k2 ${linearSearch(array, n, k2)}');
}

/// [array] is the list of data on which the linear search will perform the operation.
///
/// [n] is the size of the array.
///
/// [k] is the key that must be fetched in the array.
int linearSearch(List<int> array, int n, int k) {
  for (int i = 0; i <= n; i++) {
    if (array[i] == k) {
      return i;
    }
  }
  return -1;
}
